examples/04_pong/systems/bounce.rs

```diff
@@ -0,0 +1,78 @@
+use {Ball, Paddle, Side};
+use amethyst::audio::output::Output;
+use amethyst::ecs::{Fetch, Join, ReadStorage, System, WriteStorage};
+use amethyst::ecs::transform::LocalTransform;
+use audio::{play_bounce, Sounds};
+
+/// This system is responsible for detecing collisions between balls and
+/// paddles, as well as balls and the top and bottom edges of the arena.
+pub struct BounceSystem;
+
+impl<'s> System<'s> for BounceSystem {
+    type SystemData = (
+        WriteStorage<'s, Ball>,
+        ReadStorage<'s, Paddle>,
+        ReadStorage<'s, LocalTransform>,
+        Fetch<'s, Sounds>,
+        Fetch<'s, Option<Output>>,
+    );
+
+    fn run(&mut self, (mut balls, paddles, transforms, sounds, audio_output): Self::SystemData) {
+        // Check whether a ball collided, and bounce off accordingly.
+        //
+        // It might seem intuitive to just negate the horizontal or vertical velocity of the ball
+        // to represent a bounce, but because of the variable time frame, this could result in
+        // a ball being stuck inside the wall. This happens when the first time step is
+        // signficantly larger than the second one, as that will make the second time step
+        // unable to move the ball back out, causing its velocity to be negated again, and again...
+        for (ball, transform) in (&mut balls, &transforms).join() {
+            use ARENA_HEIGHT;
+
+            let ball_x = transform.translation[0];
+            let ball_y = transform.translation[1];
+
+            // Bounce at the top of the bottom of the arena.
+            if ball_y <= ball.radius {
```

Comment:
I would actually take this a step further and also check to see if the ball is traveling towards the boundary in question by checking the current velocity.  For example if the ball is hitting the top wall make sure it has a positive Y velocity before executing the bounce code.  The only reason this matters is to avoid multiple bounce sound effects being played per collision.